
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to HumanCircRNADataParser’s documentation! &#8212; HumanCircRNADataParser 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-humancircrnadataparser-s-documentation">
<h1>Welcome to HumanCircRNADataParser’s documentation!<a class="headerlink" href="#welcome-to-humancircrnadataparser-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>Contents:</p>
<div class="toctree-wrapper compound">
</div>
<dl class="class">
<dt id="hcrdp.DataParser">
<em class="property">class </em><code class="descclassname">hcrdp.</code><code class="descname">DataParser</code><span class="sig-paren">(</span><em>ef</em>, <em>hcf</em>, <em>mcf</em>, <em>mclf</em>, <em>hrsinef</em>, <em>mrb1b2f</em>, <em>extend_sine</em>, <em>extend_circRNA</em>, <em>extend_intron</em>, <em>comp_distance_buffer_high</em>, <em>comp_distance_buffer_low</em><span class="sig-paren">)</span><a class="headerlink" href="#hcrdp.DataParser" title="Permalink to this definition">¶</a></dt>
<dd><p>This is the main class of the tool.  From here, you can call various methods that work to convert and analyze genomic data files:
User can choose to dump all info into this class call and call each function in this order: exonToIntron(), intronExtender(), then MainParser() or call each individually with different arguments
Calling each function in order will yield the files desired if the string arguments in the constructor are all defined.
Whenever you call a function make sure that the following file names are empty or do not exist:
When calling exonToIntron: 1 file is saved: intron_file.bed
When calling intronExtender: 1 file is saved: extended_intron_file.bed
When calling mainParser: 39 files are saved: extended_lifted_mouse_circRNA_file.bed hcf_elmcf.bed, hcf_elmcf_same_start.bed, hcf_elmcf_same_end.bed, hcf_elmcfss_sine.bed, hcf_elmcfse_sine.bed, hcf_elmcfsss_unextended.bed, hcf_elmcfses_unextended.bed, hesu_nodups.bed, heeu_nodups.bed, hcb_sine.bed, hcbs_nodups.bed, hcbs_reextended.bed, hc_extended.bed, mc_same.bed, mc_same_sine.bed, mcss_nodups.bed, forced_liftover_mcss.bed, flm_start_extended.bed, flm_end_extended.bed, fse_b1b2.bed, fee_b1b2.bed, fseb_unextended.bed, feeb_unextended.bed, mcb_both.bed, mcbb_nodups.bed, introns_mcbb.bed, imcbb_unextended.bed, forced_liftover_mcf_human.bed, humanCircRNAfinalextended.bed, hcf_normal.bed, hcfn_nodups.bed, hcrpm.bed, hcrpm_nodups.bed, cofmv.bed, comhvp.bed, comparison_of_mouse_human_final.bed, cofmvv_use.bed, narrow_list_human_mouse.bed, nlhm_final.bed   
(it might be a good idea to set up a separate empty directory prior to caling these methods to contain these files)
The parameters hcf, mcf, mclf, hrsinef, and mrb1b2f, must be defined to use this code.
If you want to look at another genome, you must call the class again to redefine elements from the new genome.
The other parameters here can be defined later or redefined in calls to the functions, and the function definition of the parameters take priority.
This is made to work with Python v2.7.12.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>ef</strong> (<em>string</em>) – the name of the file containing the exons of the genome of interest in bed format</li>
<li><strong>hcf</strong> (<em>string</em>) – the name of the file containing human circular RNA data in bed format</li>
<li><strong>mcf</strong> (<em>string</em>) – the name of the file containing the circular RNA data for the genome of interest in bed format</li>
<li><strong>mclf</strong> (<em>string</em>) – the name of the file containing the circular RNA data for the genome of interest lifted over to human coordinates using the liftOver tool from the genome browser</li>
<li><strong>hrsinef</strong> (<em>string</em>) – the name of the file containing the data for the SINEs from human repeats in the genome in bed format</li>
<li><strong>mrb1b2f</strong> (<em>string</em>) – the name of the file containing the SINE equivalents contained the repeats of the genome of interest</li>
<li><strong>extend_sine</strong> (<em>int</em>) – the number of nt that is the max distance away from the circRNA for a SINE to be considered flanking to a circRNA (default 2000)</li>
<li><strong>extend_circRNA</strong> (<em>int</em>) – the max number of nucleotides that can be considered a negligible distance when determining whether the circRNA in the genome of interest is considered to be in the same position as in the human genome (default 50)</li>
<li><strong>extend_intron</strong> (<em>int</em>) – the amount of nucleotides to extend introns to act as a buffer to overlap circRNAs in order to be considered flanking to the circRNA (default 10)</li>
<li><strong>comp_distance_buffer_high</strong> (<em>int</em>) – the higher limit of how many nucleotides apart the start end coordinates of a circRNA in a given genome must be from the start and end coordinates of the other genome in order to be considered corresponding circRNAs, used in creating the nlhm_final file (default 50)</li>
<li><strong>comp_distance_buffer_low</strong> (<em>int</em>) – the lower limit for the number of nuleotides apart the start and end coordinates of a circRNA in a given genome must be from the start and end coordinates of the other genome in order to be considered corresponding circRNAs, used in creating the nlhm_final file (default -50)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="hcrdp.DataParser.exonToIntron">
<code class="descname">exonToIntron</code><span class="sig-paren">(</span><em>ef2</em><span class="sig-paren">)</span><a class="headerlink" href="#hcrdp.DataParser.exonToIntron" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an exon file to an intron file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>ef2</strong> – string of the bed file containing information on the exons of the genome of interest (default self.ef)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hcrdp.DataParser.intronExtender">
<code class="descname">intronExtender</code><span class="sig-paren">(</span><em>inf2</em><span class="sig-paren">)</span><a class="headerlink" href="#hcrdp.DataParser.intronExtender" title="Permalink to this definition">¶</a></dt>
<dd><p>This extends the intron coordinates by 10 nt in both directions</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>inf2</strong> – Optional string parameter that represents an intron file that the user may want to extend separately (default “intron_file.bed”, the file produced at the end of the last exonToIntron() call</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hcrdp.DataParser.mainParser">
<code class="descname">mainParser</code><span class="sig-paren">(</span><em>feiom</em>, <em>fextend_sine</em>, <em>fextend_circRNA</em>, <em>fextend_intron</em>, <em>fcomp_distance_buffer_high</em>, <em>fcomp_distance_buffer_low</em><span class="sig-paren">)</span><a class="headerlink" href="#hcrdp.DataParser.mainParser" title="Permalink to this definition">¶</a></dt>
<dd><p>The main function with the purpose of analyzing, comparing, and producing files with respect to the genome of interest’s relation to the human genome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>feiom</strong> (<em>string</em>) – a string representing the file containing the extended introns of the genome of interest(default eiom as definedby the class) (default2 “extended_intron_file.bed”)</li>
<li><strong>fextend_sine</strong> (<em>int</em>) – the number of nt that is the max distance away from the circRNA for a SINE to be considered flanking to a circRNA (default 2000)</li>
<li><strong>fextend_circRNA</strong> (<em>int</em>) – the max number of nucleotides that can be considered a negligible distance when determining whether the circRNA in the genome of interest is considered to be in the same position as in the human genome (default extend_circRNA)</li>
<li><strong>fextend_intron</strong> (<em>int</em>) – the amount of nucleotides to extend introns to act as a buffer to overlap circRNAs in order to be considered flanking to the circRNA (default 10)</li>
<li><strong>fcomp_distance_buffer_high</strong> (<em>int</em>) – the higher limit of how many nucleotides apart the start end coordinates of a circRNA in a given genome must be from the start and end coordinates of the other genome in order to be considered corresponding circRNAs, used in creating the nlhm_final file (default 50)</li>
<li><strong>fcomp_distance_buffer_low</strong> (<em>int</em>) – the lower limit for the number of nuleotides apart the start and end coordinates of a circRNA in a given genome must be from the start and end coordinates of the other genome in order to be considered corresponding circRNAs, used in creating the nlhm_final file (default -50)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to HumanCircRNADataParser’s documentation!</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Richard Clayton.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/index.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>